### PR TITLE
[#IOPID-231] Add dependency traces to redis requests

### PR DESCRIPTION
**Pull request by gquadrati**
#### List of Changes

- Add Redis dependency trace wrapper

#### Motivation and Context

appinsights does not trace Redis v4 calls.

#### How Has This Been Tested?

Manual tests

#### Types of changes

- [x] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/session-manager/src/utils/redis-trace-wrapper.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  functionName: string,
+  clientName: string,
+  appInsightsClient?: appInsights.TelemetryClient,
+): T {
+  return async function (...args: unknown[]) {
+    const startTime = Date.now();
+
+    try {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const result = await (originalFunction as any).apply(redisClient, args);
```
Comment (BurnedMarshal) [inline apps/session-manager/src/utils/redis-trace-wrapper.ts:19]: This works even with sync methods of the Redis client? Wrapping only the COMMANDS should be fine. I'll check if it works only with legacy mode `off`.

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: apps/session-manager/env.example [hunk 2]
```diff
@@ -7,6 +7,8 @@ APPINSIGHTS_SAMPLING_PERCENTAGE=100
 # cloud_RoleName value; if undefined, WEBSITE_SITE_NAME will be used
 APPINSIGHTS_CLOUD_ROLE_NAME="session-manager-dev"
 
+APPINSIGHTS_REDIS_TRACE_ENABLED="true"
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply apps/session-manager/env.example:10]: Not in this PR

File: apps/session-manager/src/utils/redis-trace-wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+function wrapRedisClusterClient(
+  client: redis.RedisClusterType,
+  clientName: string,
+  appInsightsClient?: appInsights.TelemetryClient,
+) {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const clientAsObject = client as Record<any, any>;
+  for (const functionName in clientAsObject) {
+    if (typeof clientAsObject[functionName] === "function") {
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply apps/session-manager/src/utils/redis-trace-wrapper.ts:59]: a list of the one used is in `__mocks__/redis.mocks.ts`

File: apps/session-manager/src/utils/redis-trace-wrapper.ts (hunk 1 above)
Comment (gquadrati) [reply apps/session-manager/src/utils/redis-trace-wrapper.ts:19]: await a non-promise function is fine. But do you know how to filter commands only?

Comment (BurnedMarshal) [reply apps/session-manager/src/utils/redis-trace-wrapper.ts:19]: We could use this code:
```typescript
import commands from "@redis/client/dist/lib/cluster/commands";

...

Object.keys(commands) // List of commands
```

**Review by BurnedMarshal — APPROVED** [review]

**Pull request merged**